### ISSUE TITLE
Expose heartbeat?

### ISSUE THREAD

**Issue by michaelsauter** [zendesk/ruby-kafka#355]
I have a consumer which reads messages in batches, and needs to call an external API a few times for each batch (there is a limit on the external API, and the Kafka batch size might exceed that).

In that case, I've seen the consumer session timeout and when the consumer tries to commit the offsets, the broker thinks that the consumer died already.

I see a couple of options here:

1. Reduce the amount of bytes to fetch from a partition
2. Increase the session timeout
3. Send a heartbeat while processing the batch

My question if it makes sense to do (3) ? I have already done (1), and that helps, but potentially the external API is slow and/or rate limited so I don't really know how long each API call can take so I might still have to do a heartbeat. I could do (2) but that would then increase the time until Kafka detects a consumer died. (3) sounds good to me because I can keep the timeout short, and in my code I know if I'm still alive, just not done yet.

Does that make sense? If so, should it be possible to do e.g. `consumer.send_heartbeat_if_necessary`? I experimented with this by doing `consumer.instance_variable_get("@heartbeat").send_if_necessary`, and that has the desired effect.

What do you think?

**Comment by dasch** [comment]
What does your processing code look like? Is your API call synchronous?

There's a risk in decoupling the heartbeat from the actual processing, in that you may end up with a zombie consumer holding on to partitions. That's why I'm wary of exposing the heartbeat trigger directly, as well as why it doesn't happen on a background thread.

**Comment by michaelsauter** [comment]
Yes, the API calls are asynchronous :/

Good point about the zombie processes. However, I agree only partially.

It makes a lot of sense not to trigger heartbeats automatically in a background thread because a) it might not be obvious that this is happening and b) as a user I cannot control it so in the end it is likely that some users would run into problems. 

But exposing an API to trigger a heartbeat on demand is a different matter I think. It is completely obvious (because it's an explicit opt-in) and as a user, I have control over it (e.g. I could have some check to allow only 3 manual heartbeats before giving up).

In our case, there is actually just one consumer in the group right now, and I know that doing one API call while processing a batch is completely fine. However, there are situations where I might need to do  more API calls, and then I want to heartbeat after each API call.
If the API I'm talking to is really really slow, it's okay to timeout though and process the messages again, that is actually desired as I want to basically spin until the API is back up.
But I don't want to spin forever because I get a batch which triggers more API calls then what the session timeout will allow me to do.

So ... having an official way to trigger a heartbeat would be really nice! :) Happy to get a PR done if you decide to accept this feature request.

**Comment by dasch** [comment]
Hmm. I'm still not sure it's worth the complexity cost. Have you tried
running with a longer session timeout? In most cases, consumers will send a
leave message on shutdown, so I don't think you'll run into problems very
often.

**Comment by michaelsauter** [comment]
No, I did not try to increase the session timeout. I would like to avoid that. I still think it is much better to have a shorter timeout, but telling the broker more often that we're still alive. Also, a longer timeout will make problems less likely, but it is not a reliable solution.

Let me try to explain our situation in more detail. The consumer in question is reading from one topic, and each message in the topic is some kind of change to a resource. The consumer now needs to sync this resource with an external API. However, the endpoint we're calling is not always the same. In one case it is a "batch update" (limit of 100 resources per batch), in another it is a "delete" (no batching possible).

To illustrate: I have reduced the amount of bytes to fetch to a value that ensures we're receiving at most 100 messages per fetch. Now, assume we get 100 messages in a batch, but of those messages, some are concerning the same resource, so we are left with say 60 "resource changes". Of those changes, 55 need to be send to the "batch update", and 5 need to be send to the "delete" endpoint. This means that we'll have 6 API calls.

It's easy to see that the number of API calls is very hard to predict, and therefore choosing a reliable session timeout is not possible. At the same time, reducing the number of bytes to fetch will certainly help, but it also lowers performance considerably as we can't send lots of changes to the batch update at once.

The only reliable way to get out of this dilemma, I think, is to send a heartbeat manually. I know that one API call is definitely within the session timeout, so the solution is to send a heartbeat after each API call if more than `heartbeat_interval` elapsed. Doing so seems to be completely safe to me, there is no chance that the consumer will hold on to a partition in error.

I'm also not quite convinced that this introduces complexity that isn't there already. As I mentioned, for now I'm using `consumer.instance_variable_get("@heartbeat").send_if_necessary`, so a feature that already exists, but that is not exposed. Exposing it certainly is something that I understand you're not doing lightly, but I don't see the drawback. You already have the functionality, and you already are sending the heartbeat at unpredictable intervals (which are at least 5s apart, okay, but that would also be the case once exposed).

Hope that helps in understanding our issue a bit better. Would be awesome if you consider this again.

I'm not quite sure what you meant with

> In most cases, consumers will send a leave message on shutdown, so I don't think you'll run into problems very often.

How does this help with my situation, or am I completely missing the point? ;)

**Comment by dasch** [comment]
Well, the session timeout is really only used to ensure that clients that
crash without sending a leave message get pruned from the group and their
partitions reassigned - so a longer timeout will only affect those cases.

I'm on vacation at the moment, but I'll take a look at this when I get home.

**Comment by dasch** [comment]
I would be open to a PR that implements `Consumer#heartbeat`.

**Comment by michaelsauter** [comment]
Awesome! I'm away at the moment, but @headshota, could you look into opening a PR if you get the time?

**Comment by tylerholien** [comment]
FWIW the official Java client does heartbeating in a background thread: https://kafka.apache.org/0101/documentation.html#upgrade_1010_notable

**Comment by dasch** [comment]
@tylerholien it sounds like the entire consumer join/fetch/leave flow is being done in a separate thread, probably using an inter-thread communication mechanism when the user code calls `poll`. This is something I've thought about, but it would require quite a bit of work to make it 100% reliable. Error handling is always tricky when you introduce multithreading, for instance. I'm not sure it's the right time to start working on that.

Note that their implementation doesn't actually "do" heartbeating in a background thread: the heartbeat is still invoked only after the user code calls `poll` in the foreground thread. It's just the IO work that's being done in the background.

**Comment by tylerholien** [comment]
It is certainly complex and challenging! Sorry, I didn't mean to imply that it was trivial or that you should do it now. As of 0.10.1 the consumer does do heartbeating in a background thread (`HeartbeatThread` in [AbstractConsumer.java](https://github.com/apache/kafka/blob/0.10.2/clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java)). This gives us the ability to have long times between polls and still remove dead group members relatively quickly. Easier said than done, though! Thanks for a great library!

**Comment by dasch** [comment]
My reading of https://github.com/apache/kafka/blob/0.10.2/clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java#L269-L277 is that while a background thread _manages_ the heartbeat, the foreground threads needs to manually indicate that it's still running – so it's, functionally speaking, still a sequential operation relative to the message processing. It's just that from the point of view of the foreground thread, the heartbeat has virtually no latency (since the hard work happens in the background thread.)

**Comment by dasch** [comment]
✅

**Comment by tylerholien** [comment]
It does do the heartbeat in the background without calling any methods on the consumer. You can confirm this by sleeping for longer than the session timeout:

        Properties props = new Properties();
        props.put("bootstrap.servers", "localhost:9092");
        props.put("enable.auto.commit", "false");
        props.put("session.timeout.ms", "10000"); // the default in 0.10.1 and later
        props.put("key.deserializer", "org.apache.kafka.common.serialization.StringDeserializer");
        props.put("value.deserializer", "org.apache.kafka.common.serialization.StringDeserializer");
        KafkaConsumer<String, String> consumer = new KafkaConsumer<>(props);
        consumer.subscribe(Arrays.asList("test_topic"));
        while (true) {
            ConsumerRecords<String, String> records = consumer.poll(100);
            System.out.println(String.format("Got %d records. Sleeping.", records.count()));
            try { Thread.sleep(60000); } catch(InterruptedException e){}
            System.out.println("Slept. Committing.");
            consumer.commitSync();
            System.out.println("Committed.");
        }

You will get an error with consumer versions prior to 0.10.1. As of 0.10.1, this works just fine. You can also watch network traffic and see the heartbeats while the above code is sleeping.

**Comment by dasch** [comment]
@tylerholien hmm... that's not my reading of the documentation, and also doesn't really make any sense. Try sleeping for a _long_ time and see when the heartbeat thread triggers a group exit.

If your processing code is permanently blocked on IO you need the heartbeat to stop – otherwise there's no point in even having a heartbeat.

**Comment by tylerholien** [comment]
Yeah, the documentation about these timeouts is pretty poor. The important bits are hidden in the [notable changes in 0.10.1.0](https://kafka.apache.org/0101/documentation.html#upgrade_1010_notable) section, and even if you find them they're still a little confusing in my opinion.

This part covers the permanently blocked on IO issue:

> There is a new configuration `max.poll.interval.ms` which controls the maximum time between poll invocations before the consumer will proactively leave the group (5 minutes by default).

So, if you sleep for longer than `max.poll.interval.ms`, you still get kicked out of the group and the commit will fail. However, `max.poll.interval.ms` can be _much_ longer than `session.timeout.ms` because the heartbeating is happening while you are sleeping. That way, if the consumer is no longer sending heartbeats (e.g. because it was `kill -9`ed, it lost network access, etc.), the group will rebalance within the shorter `session.timeout.ms` (which defaults to 10 seconds as of 0.10.1).

Here's the whole paragraph about these changes (as of 0.10.1) found in the [notable changes in 0.10.1.0](https://kafka.apache.org/0101/documentation.html#upgrade_1010_notable) section:

> The new Java Consumer now supports heartbeating from a background thread. There is a new configuration `max.poll.interval.ms` which controls the maximum time between poll invocations before the consumer will proactively leave the group (5 minutes by default). The value of the configuration `request.timeout.ms` must always be larger than `max.poll.interval.ms` because this is the maximum time that a JoinGroup request can block on the server while the consumer is rebalancing, so we have changed its default value to just above 5 minutes. Finally, the default value of `session.timeout.ms` has been adjusted down to 10 seconds, and the default value of `max.poll.records` has been changed to 500.

Here is updated code that fails because of the `max.poll.interval.ms`:

```java
Properties props = new Properties();
props.put("bootstrap.servers", "localhost:9092");
props.put("group.id", "timeout-test");
props.put("enable.auto.commit", "false");
props.put("session.timeout.ms", "10000"); // the default
props.put("max.poll.interval.ms", "20000");
props.put("request.timeout.ms", "25000");
props.put("key.deserializer", "org.apache.kafka.common.serialization.StringDeserializer");
props.put("value.deserializer", "org.apache.kafka.common.serialization.StringDeserializer");
KafkaConsumer<String, String> consumer = new KafkaConsumer<>(props);
consumer.subscribe(Arrays.asList("test_topic"));
while (true) {
    ConsumerRecords<String, String> records = consumer.poll(100);
    System.out.println(String.format("Got %d records. Sleeping.", records.count()));
    try { Thread.sleep(60000); } catch(InterruptedException e){}
    System.out.println("Slept. Committing.");
    consumer.commitSync();
    System.out.println("Committed.");
}
```

**Comment by dasch** [comment]
That does sound like an improvement, although it'll be more work to implement. I'll be revisiting the entire Consumer design in the near future and will be sure to look at Heartbeat while I'm at it.
